Add support for object-persistor

gh2k opened the pull request:
### Description

This adds support for https://github.com/overleaf/object-persistor to Docstore, in place of direct S3 support.

The purpose of this is to allow us to store data on GCS as well as S3, with a fallback so that we can live-migrate without missing any data.

Sorry about this...
![image](https://user-images.githubusercontent.com/6530014/88380651-d75cd900-cd9c-11ea-99dd-adb17d32cb4e.png)

#### Related Issues / PRs

This is the main part of the work for overleaf/issues#3107 - after which we can migrate the data

### Review

Ths bulk of the work is in https://github.com/overleaf/docstore/commit/2d285e2cdd69af11b6a8bc40b9228e289b577838 - Unfortunately, I've promisified and implemented in one single commit. I did consider doing this in multiple, but given the amount of code that there was to change it made more sense to treat this as a rewrite.

As such, I think it makes sense to view that particular patch side-by-side with the original as it's hard to read the diff otherwise. - The same for the unit tests for the file as well.

The acceptance tests now use `fake-gcs` as the backend, instead of talking directly to S3 with some live credentials. I have not added additional acceptance tests for using `@overleaf/object-persistor` with S3, or with a fallback. These are very comprehensively as part of the https://github.com/overleaf/filestore acceptance tests, so I don't think it makes sense to reinvent the wheel here as well.

I updated all modules to their latest minor release. Most of these are dev dependencies.

The code references S3 throughout. I haven't changed this, to keep the patch size from becoming even more unmanageable.

There are two additional npm modules.

* `tiny-async-pool` adds a promisey equivalent of `async.eachLimit`
* `streamifier` is a convenience wrapper to download a stream into a string

I've looked at the code for both, and they look fine. We've used `tiny-async-pool` in FIlestore, and `streamifier` in the Filestore acceptance tests.

Requesting a double-review due to scope. Apologies for the size of the PR - I would be happy to talk it through on a call, which can help for something of this size.

#### Potential Impact

This could break the whole thing, which would be very bad.

#### Manual Testing Performed

Currently, I've only tested that the tests pass and that my dev environment isn't broken. I still need to:

- [x] Use curl and editing projects to [un]archive docs into fake-gcs
- [x] Try using the dev env to archive into s3mock as well, to be sure that this works

In staging, we will need to:

- [x] Archive some project(s) to S3 (via curl)
- [x] Unarchive them by editing the project

Later, after this version is in production, we will need to:

- [x] Set up staging to use gcs with S3 as a fallback
- [x] Check that it works as above
- [ ] ~Check with copy-on-miss~ **Something about the workflow breaks copy-on-miss**. I haven't had time to look into it, but I think we're probably best just not using it.

### Deployment

See checklist below.

One thing that I am anxious about however is that we are actually deleting stuff already. When projects are hard-deleted we call Docstore to delete any docs. This could potentially leave us with a condition where a user deletes something which we end up resurrecting again if it's being copied. - I don't *think* this will be a problem, because:

a) We delete data on both the primary and fallback when something is deleted, so there's not a high chance that the object will be there for copying at the time the transfer appliance wants to copy the file, and
b) We plan to go through and delete any cruft related to projects that have already been deleted afterwards, anyway (we only delete when expiring a hard-deleted project, so there will be old stuff still around)

... as such, I think this is safe but would appreciate input from @jdleesmiller on this in case I am missing something.

#### Deployment Checklist

In production, we will need to:

- [ ] Deploy with a canary
- [ ] Wait until a number of projects have been archived/unarchived without error
- [ ] Deploy
- [ ] Wait for a bit
- [ ] Deploy with GCS and S3 as a fallback
- [ ] Wait for a bit
- [ ] Copy the data from S3

I do not recommend a canary deploy with staging, as there will be a limited number of projects that we can work with and we don't want the requests to go to a different pod when we are testing.

#### Metrics and Monitoring

Once deployed, we will want to ensure that the hit and error rates are approximately the same, and keep an eye on the timings.

It might be helpful to add metrics on archiveDoc and unarchiveDoc to see that these things are happening.

#### Who Needs to Know?

cc @jdleesmiller @emcsween

gh2k commented:
I've given this a pretty good thrashing in my dev env, and I'm happy with it. It requires overleaf/dev-environment#376 in order to work - otherwise docstore will crash on startup as there isn't a backend configured for `object-persistor` to work with.

ShaneKilkelly commented on app/js/DocArchiveManager.js:
```diff
@@ -1,295 +1,179 @@
+  } else if (Array.isArray(doc)) {
+    mongoDoc.lines = doc
+  } else {
+    throw new Error("I don't understand the doc format in s3")
```
Thought: are we using OError in this project? Would we be able to debug this issue if if popped up in production?

gh2k replied:
I don't think we were, but it's being pulled in via `object-persistor` so I think it would be better to use it, yes. :+1:

ShaneKilkelly commented on app/js/DocArchiveManager.js:
```diff
@@ -1,295 +1,179 @@
+    'removing doc from mongo and persistor'
+  )
+  const doc = await promisify(MongoManager.findDoc)(projectId, docId, {
+    inS3: 1
```
I assume `inS3` is a legacy flag, and that it really means `inStorage` with this new object-persistor system? (Rather than being specifically in S3)

gh2k replied:
Yeah, I thought it better to leave the names of things intact (at least for now) to cut down on blast radius as much as possible.

warncke commented on test/acceptance/js/helpers/DocstoreApp.js:
```diff
@@ -15,6 +15,13 @@ const app = require('../../../../app')
 const settings = require('settings-sharelatex')
 
+// treat unhandled promise rejections as failures
+process.on('unhandledRejection', (e) => {
```
in 10.17+ this can be done with https://nodejs.org/docs/latest-v10.x/api/cli.html#cli_unhandled_rejections_mode

gh2k replied:
Oh, that's a nice feature. I wonder if it's worth adding to the buildscripts in that case?

gh2k commented on app/js/DocArchiveManager.js:
```diff
@@ -1,295 +1,183 @@
-    })
-  },
+async function archiveAllDocs(projectId) {
+  const docs = await promisify(MongoManager.getProjectsDocs)(
```
Yes, that would be better! :+1:

gh2k commented on app/js/DocArchiveManager.js:
```diff
@@ -1,295 +1,183 @@
 const RangeManager = require('./RangeManager')
-const thirtySeconds = 30 * 1000
+const PersistorManager = require('./PersistorManager')
+const AsyncPool = require('tiny-async-pool')
```
Yeah, I think that's a good idea. I've flip-flopped around on how to CapitaliseWhenImporting, as there's a part of me that wants to be explicit about things that have been imported. I think I saw it somewhere and liked it, but I don't think it's consistent with what we usually do.

gh2k commented on app/js/DocArchiveManager.js:
```diff
@@ -1,295 +1,183 @@
-    })
-  },
+async function archiveDoc(projectId, doc) {
+  logger.log(
```
Yeah, I think so. I've tried to keep the logging consistent with the previous behaviour, again to try and reduce the amount of stuff that's changed.

gh2k commented on test/acceptance/deps/healthcheck.sh:
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
+
+# health check to allow 404 status code as valid
```
We want to know that the service is up and receiving http connections, but we don't necessarily want to know that it has a specific piece of data in it. So, pinging it and getting a 404 back is a good chance that it's working.

gh2k commented on app/js/DocArchiveManager.js:
```diff
@@ -1,295 +1,183 @@
-  },
+  await AsyncPool(
+    PARALLEL_JOBS,
+    docs.filter((doc) => !doc.inS3),
```
Yes, it should. :+1: Good catch.

warncke replied:
are you going to change this still?

gh2k replied:
Sorry, it fell into the "hidden conversations" section. Yes, need to do this.

gh2k commented on app/js/DocArchiveManager.js:
```diff
@@ -1,295 +1,183 @@
+    throw error
+  }
+
+  const md5 = crypto.createHash('md5').update(json).digest('hex')
```
The reason why we don't want it to work that way is becase the persistor itself doesn't know where the stream is coming from. The persistor logic goes like this:

- Did the client supply an md5 hash? If so, verify that it matches the data on the backend when it is uploaded.
- Otherwise, compute the md5 hash of the data as it passes through and verify that the remote has the same md5 after it is uploaded.

This is subtly different, as if we know the md5 beforehand we can use the S3/GCS native implementation that validates the data server-side against the supplied hash and fail if it doesn't.

If we don't know it in advance, we can still validate it as far as the data is concerned from object-persistor's perspective, but we don't want to consume all of the data first to calculate the size as it might be quite large, so we calculate as we go, verify the upload afterwards and delete it if it's not right. That is a more expensive way of doing it.

The other advantage of being able to supply the md5 in advance is that object-persistor doesn't know where the stream is coming from. It could be passing through something from a downstream http client. If that client can tell us what the md5 is beforehand then we can use it to validate both the upload to the storage engine (S3/GCS) and the upload from the client to object-persistor as the middle-man.

Given that we've already serialised the content to json here and it shouldn't be huge, I think it makes sense to calculate it here instead.

gh2k commented:
Thanks for the review - good points. With regard to adding OError3 and cleaning up the logs vs. leaving the logging as-is? Might it be better to leave this as-is for now open a ticket to sort this out in a follow-up? I think it might be better to tackle the whole of docstore in one go, and leave the behaviour unchanged for now so that the logs look the same as before when we deploy this.

I'm open to changing it in the archive manager now, but it might be better to go through the whole of docstore at once as a follow-up to get the whole thing cleaned up together. WDYT, @ShaneKilkelly @warncke ?

gh2k commented:
Updates done. I've removed the unhandled promise rejection handler as it was added to the buildscripts in overleaf/dev-environment#378 so I ran this instead.

We don't have util from web, so there is no promisifyAll. I've modified the `timeAsyncMethod` shim to add the promisification into there for `MongoManager`.

gh2k commented:
Added a commit to fall back to the old `AWS_BUCKET` variable so that this can be tested in staging without any config changes.

gh2k commented:
Tested on staging. These things work:
* Archive to GCS
* Unarchive from GCS
* Unarchive from S3 (as fallback)

Also, verified that we're OK with:

* Archive to S3 (as primary)
* Unarchive from S3 (as primary)

**Copy on miss does *not* work**. Unless anyone wildly disagrees I think we should just not use it. There's something convoluted about the way it's trying to get an md5 hash before copy.

gh2k commented:
WRT copying the data, there are two ways we could go...
* Use the transfer appliance
* Go through all projects with docs in S3 and archive/unarchive them

I *think* it should be safe to use the transfer appliance. The Bad Things that could happen would be that it resurrects a file that shouldn't be there, or it overwrites a file with an older version.
* For the latter, the event-based hold will prevent that (and log an error)
* For the former, I think it's OK because we delete things with the persistor in both places at once, so the copy process should come across a missing file and skip it (and log an error)

The difference between this and when we did it with Filestore is that we are already deleting stuff from S3/GCS for Docstore, both when projects are unarchived and when they are hard-deleted.

The advantage of doing this with a script to archive/unarchive things would be that we sort out the GDPR problem at the same time. What would happen is:
* Find projects in Mongo with docs that are `inS3` set, and where their archived date is less than the switch-to-GCS date
* Unarchive the project. Data is removed from S3 and added to Mongo
* Archive the project. Data is archived to GCS and removed from Mongo.

This way, we only archive stuff which we already have in our database. Anything left on S3 at the end should be for projects that no longer exist. - The downside is that it costs us more in CPU and will be slower. Can you fragment Mongo collections? That could also be bad.

WDYT, cc @jdleesmiller

jdleesmiller commented:
Re not using copy on miss: that sounds fine to me.

Re script vs transfer tool: some further thoughts below.

Script:

- I like the script idea --- nice that it solves both the GDPR and transfer problems, it only transfers what we need, and it hopefully can reuse existing mechanisms for locking to make the process safe.
- I feel like we might encounter more historical oddities, e.g. doc lines with `\r`s, that might cause problems with running the script, but of course that is an 'unknown unknown'.
- We'd have to be careful that the script didn't have user-visible effects, e.g. old projects appearing to have been modified and jumping back to the top of the dash, because that sort of thing generates a lot of support.
- If the script was written to do projects in parallel, it might not be that much slower.

Transfer appliance:

- If we were to stop archiving for a bit (presumably this transfer would only take a few days), that might provide some additional assurance about consistency. (Can we do that?) We can't stop unarchiving, but at least there would not be any new versions to overwrite older versions with, so the only problem we'd have to worry about would be deleting an object from GCS then copying it from S3 before it was deleted from S3 (a narrow window, possibly no window).
- I guess we're still going to have to write a script in this case to look for orphaned objects (GDPR), and that script could probably include a check for objects with docs that are in mongo and not `inS3`, which might be due to the bug above or some previous bugs, as well as docs that no longer exist in mongo.

So, I am not leaning strongly one way or the other at this point. Do you have a leaning?

gh2k commented:
>     * I feel like we might encounter more historical oddities, e.g. doc lines with `\r`s, that might cause problems with running the script, but of course that is an 'unknown unknown'.

Good point. From my understanding of the code, we will be splitting the data on `\n` and putting it into the Mongo lines array otherwise-unmodified, so from this perspective I think it is safe.

In theory, we could use this as an opportunity to strip out `\r`s (and null bytes, I think these have caused a problem before?) which is something we've wanted to do in the past, anyway... But I think it might be more trouble than it is worth given that we would have to play nice with the history service. (cc @briangough to confirm that sanitising the data is a bad idea)

>     * We'd have to be careful that the script didn't have user-visible effects, e.g. old projects appearing to have been modified and jumping back to the top of the dash, because that sort of thing generates a lot of support.

I don't think it will if we're calling the docstore endpoints directly rather than going via web, which is what I think we should be doing. I'll double-check, but I don't think there is code in docstore itself to bump the modification date. Archiving via the docstore API and then opening a project in web but not modifying it does not cause it to jump to the top of the list.

>     * If the script was written to do projects in parallel, it might not be that much slower.

I think it would be slower per-operation, because we have to talk to Mongo. We can still parallelise, but we have more opportunity to DOS ourselves.

> Transfer appliance:
> 
>     * If we were to stop archiving for a bit (presumably this transfer would only take a few days), that might provide some additional assurance about consistency. (Can we do that?) We can't stop unarchiving, but at least there would not be any new versions to overwrite older versions with, so the only problem we'd have to worry about would be deleting an object from GCS then copying it from S3 before it was deleted from S3 (a narrow window, possibly no window).

I don't think this is necessary because we will be archiving to GCS at this point, and if we're archiving then there shouldn't be anything there to copy from S3 and so we won't resurrect the files.

>     * I guess we're still going to have to write a script in this case to look for orphaned objects (GDPR), and that script could probably include a check for objects with docs that are in mongo and not `inS3`, which might be due to the bug above or some previous bugs, as well as docs that no longer exist in mongo.

We will, yes. I think there are both docs in Mongo for which there is no project, and files on S3 for which there is no doc. The former is easy enough to check for just in Mongo so I'm not so worried about that.

One other thing to consider is that we'd need to use docstore to delete the files that we don't want anyway if we're doing this post-transfer, as we will need to work around the administrative lock.

> So, I am not leaning strongly one way or the other at this point. Do you have a leaning?

I think slightly scriptwardly at this point. One thing I like about that approach is that we clean up the data as we go, which makes it much easier to validate the stuff that's left after the copy is finished.

jdleesmiller commented:
Thanks, good points.

> could use this as an opportunity to strip out \rs (and null bytes

I also thought that... but it seems like it might significantly expand the scope. I think the main concern is that it might go out of sync with the history. Maybe we could just record anomalies without attempting to fix them --- we'd then have a much better picture than we do now of how prevalent these problems are.

> slightly scriptwardly at this point

Same here.

gh2k commented:
> record anomalies

Good call. :+1: 

> scriptwardly

OK. I'll open a ticket for the script.

das7pad commented:
> > could use this as an opportunity to strip out \rs (and null bytes
> 
> I also thought that... but it seems like it might significantly expand the scope. I think the main concern is that it might go out of sync with the history.

The frontend runs a sanity check on carriage returns when opening a project and removes any matches in a history preserving way: https://github.com/overleaf/web/commit/fae28a8a37760074d8bc642a8199b3559d0adfce

jdleesmiller commented:
> frontend runs a sanity check

Yep, not forgotten, thanks! In the end we decided to just record the anomalies so we can get a picture of how much historical bad data we have.
